include/zephyr/drivers/bluetooth.h

```diff
@@ -0,0 +1,253 @@
+
+#define BT_DT_HCI_NAME_GET(node_id) DT_PROP_OR(node_id, bt_hci_name, "HCI")
+#define BT_DT_HCI_NAME_INST_GET(inst) BT_DT_HCI_NAME_GET(DT_DRV_INST(inst))
+
+#define BT_DT_HCI_BUS_GET(node_id) DT_STRING_TOKEN_OR(node_id, bt_hci_bus, BT_HCI_BUS_VIRTUAL)
+#define BT_DT_HCI_BUS_INST_GET(inst) BT_DT_HCI_BUS_GET(DT_DRV_INST(inst))
+
+typedef int (*bt_hci_recv_t)(const struct device *dev, struct net_buf *buf);
+
+__subsystem struct bt_hci_driver_api {
+	int (*open)(const struct device *dev, bt_hci_recv_t recv);
+	int (*close)(const struct device *dev);
+	int (*send)(const struct device *dev, struct net_buf *buf);
+#if defined(CONFIG_BT_HCI_SETUP)
+	int (*setup)(const struct device *dev,
+		     const struct bt_hci_setup_params *param);
+#endif /* defined(CONFIG_BT_HCI_SETUP) */
+};
+
+/**
+ * @brief Open the HCI transport.
+ *
+ * Opens the HCI transport for operation. This function must not
+ * return until the transport is ready for operation, meaning it
+ * is safe to start calling the send() handler.
+ *
+ * @param dev  HCI device
+ * @param recv Callback to receive data from the controller
+ *
+ * @return 0 on success or negative error number on failure.
+ */
+static inline int bt_hci_open(const struct device *dev, bt_hci_recv_t recv)
+{
+	const struct bt_hci_driver_api *api = (const struct bt_hci_driver_api *)dev->api;
+
+	return api->open(dev, recv);
+}
+
+/**
+ * @brief Close the HCI transport.
+ *
+ * Closes the HCI transport. This function must not return until the
+ * transport is closed.
+ *
+ * @param dev HCI device
+ *
+ * @return 0 on success or negative error number on failure.
+ */
+static inline int bt_hci_close(const struct device *dev)
+{
+	const struct bt_hci_driver_api *api = (const struct bt_hci_driver_api *)dev->api;
+
+	if (api->close == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->close(dev);
+}
+
+/**
+ * @brief Send HCI buffer to controller.
+ *
+ * Send an HCI command or ACL data to the controller. The exact
+ * type of the data can be checked with the help of bt_buf_get_type().
+ *
+ * @note This function must only be called from a cooperative thread.
+ *
+ * @param dev HCI device
+ * @param buf Buffer containing data to be sent to the controller.
+ *
+ * @return 0 on success or negative error number on failure.
+ */
+static inline int bt_hci_send(const struct device *dev, struct net_buf *buf)
+{
+	const struct bt_hci_driver_api *api = (const struct bt_hci_driver_api *)dev->api;
+
+	return api->send(dev, buf);
+}
+
+#if defined(CONFIG_BT_HCI_SETUP) || defined(__DOXYGEN__)
+/**
+ * @brief HCI vendor-specific setup
+ *
+ * Executes vendor-specific commands sequence to initialize
+ * BT Controller before BT Host executes Reset sequence.
+ *
+ * @note @kconfig{CONFIG_BT_HCI_SETUP} must be selected for this
+ * field to be available.
+ *
+ * @return 0 on success or negative error number on failure.
+ */
+static inline int bt_hci_setup(const struct device *dev, struct bt_hci_setup_params *params)
```

Comment:
How do we envision that this API is extended for future use cases?
Right now it is only used to set public address, but later it may be used to configure power mode or something else.
Will there be cases where some parameters are used by some vendors only?

As the API definition and driver code are located in the same place we know that these will always be in sync, so it may not be a big issue. If there is something vendor specific we can hide it behind KConfig options

Reply:
> How do we envision that this API is extended for future use cases?
> Right now it is only used to set public address, but later it may be used to configure power mode or something else.
> Will there be cases where some parameters are used by some vendors only?

All good questions. The parameter is in the form of a struct in order to provide some more flexibility and easier extension in the future. I don't see how this could contain vendor-specific extensions though, since our host-side code is (at least right now) rather vendor agnostic.

In general, discussing future of this API is something we should certainly do, but I feel that it's out of scope of this PR. The size of this PR is already quite large as-is, so I've tried to keep the semantics as close to 1:1 to the old API as possible, and leave any further API redesign for later.